lib/widgets/compose_box.dart

```diff
@@ -157,20 +157,80 @@ class ComposeTopicController extends ComposeController<TopicValidationError> {
+    }
+
+    return trimmed;
   }
 
   /// Whether [textNormalized] would fail a mandatory-topics check
   /// (see [mandatory]).
   ///
   /// The term "Vacuous" draws distinction from [String.isEmpty], in the sense
   /// that certain strings are not empty but also indicate the absence of a topic.
-  bool get isTopicVacuous => textNormalized == kNoTopicTopic;
+  bool get _isTopicVacuous {
+    bool result = textNormalized.isEmpty
+      // We keep checking for '(no topic)' regardless of the feature level
+      // because it remains equivalent to an empty topic even when FL >= 334.
+      // This can change in the future:
+      //   https://chat.zulip.org/#narrow/channel/412-api-documentation/topic/.28realm_.29mandatory_topics.20behavior/near/2062391
+      || textNormalized == kNoTopicTopic;
+
+    // TODO(server-10): simplify
+    if (store.zulipFeatureLevel >= 334) {
+      result |= textNormalized == store.realmEmptyTopicDisplayName;
+    }
 
-  @override
-  List<TopicValidationError> _computeValidationErrors() {
+    return result;
+  }
+
+  /// The send destination as a string.
+  ///
+  /// If topics are [mandatory], returns a string formatted with a topic name
+  /// when the trimmed input is non-empty. E.g.: "#stream name > topic name".
+  ///
+  /// If topics are not [mandatory], returns a string formatted with a topic
+  /// name only when the user intentionally leave the trimmed input empty.
+  ///
+  /// Otherwise, returns a string formatted without a topic name.
+  /// E.g.: "#stream name".
+  // No i18n of the use of "#" and ">" strings; those are part of how
+  // Zulip expresses channels and topics, not any normal English punctuation,
+  // so don't make sense to translate. See:
+  //   https://github.com/zulip/zulip-flutter/pull/1148#discussion_r1941990585
+  String getDestinationString({
+    required String streamName,
+    required bool contentHasFocus,
+  }) {
+    final textTrimmed = text.trim();
+    if (textTrimmed.isNotEmpty) {
+      return '#$streamName > $textTrimmed';
+    }
+
+    assert(_isTopicVacuous);
+    if (
+      // Sending to a vacuous topic (see [_isTopicVacuous]) is not possible if
+      // topics are [mandatory].
+      mandatory
+      // Do not fall back to a vacuous topic unless the user explicitly chooses
+      // to do so (by skipping topic input and moving focus to content input),
+      // because we expect a call to action for the user to pick one first.
+      || !contentHasFocus
+    ) {
+      return '#$streamName';
+    }
+
+    final vacuousTopicDisplayName =
+      // TODO(server-10): simplify away conditional
+      textNormalized.isEmpty ? store.realmEmptyTopicDisplayName : textNormalized;
+    return '#$streamName > $vacuousTopicDisplayName';
+  }
+
+  @override List<TopicValidationError> _computeValidationErrors() {
     return [
-      if (mandatory && isTopicVacuous)
+      if (mandatory && _isTopicVacuous)
         TopicValidationError.mandatoryButEmpty,
```

Comment:
Looking again at this, it looks like `TopicValidationError.mandatoryButEmpty` should really be called `TopicValidationError.mandatoryButVacuous`, is that right?

Reply:
Perhaps similarly for `zulipLocalizations.topicValidationErrorMandatoryButEmpty`?

Reply:
I think it makes sense to rename the enum value, because the distinction between empty and vacuous matters to the developer. Not too sure about translator/user-facing strings, but I understand the benefit of having variable names and string names match.